### PR TITLE
Fix randomly failing tests

### LINKED ISSUE #270: Random seed related test failure
On [41dae34](https://github.com/astro-informatics/sopt/pull/265/commits/41dae3406fe8b6d47f16e801b2293ecefdecb57e) test #8 failed on my Mac on the first time I ran it. All following runs passed so I could not reproduce. Could this be caused by the random seed? @MatthijsMars @CosmoMatt 

```ctest
8/39 Test  #8: test_sdmm ............................................***Failed    0.49 sec
[2022-03-18 11:12:38.174] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.175] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.203] [sopt] [critical] Performing SDMM 
[2022-03-18 11:12:38.244] [sopt] [critical] Performing SDMM 

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_sdmm is a Catch v2.13.7 host application.
Run with -? for options

Randomness seeded to: 1647601958

-------------------------------------------------------------------------------
SDMM with ||x - x0||_2 functions
  With different L
-------------------------------------------------------------------------------
/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:211
...............................................................................

/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:229: FAILED:
  CHECK( func(result) <= func(result + epsilon) )
with expansion:
  3.7534305808 <= 3.7534304842
with messages:
  result.transpose() :=    -1.647  0.311289   0.87681 -0.863347
  epsilon.transpose() :=      0 0.0001      0      0

/Users/tkoskela/git_repos/sopt/cpp/tests/sdmm.cc:230: FAILED:
  CHECK( func(result) <= func(result - epsilon) )
with expansion:
  3.7534305808 <= 3.7534304071
with messages:
  result.transpose() :=    -1.647  0.311289   0.87681 -0.863347
  epsilon.transpose() :=      0      0      0 0.0001

===============================================================================
test cases:  3 |  2 passed | 1 failed
assertions: 85 | 83 passed | 2 failed
```

_Originally posted by @tkoskela in https://github.com/astro-informatics/sopt/issues/265#issuecomment-1072323584_

- [ ] TODO: Reproduce the failure(s) by inserting the random seed into the tests

### PR DESCRIPTION
**Pull request by SJaffa**
List of tests that fail intermittently:

- [ ] test_sdmm ([this issue](https://github.com/astro-informatics/sopt/issues/270), on Tuomas's Mac, and [on GH actions](https://github.com/astro-informatics/sopt/runs/5883416016?check_suite_focus=true)) - failing seed 1647601958 (Tuomas's mac and 1649413076 (ubuntu-20.04-g++-9-mpi:off-openmp:on ) 
- [ ] [test_power_method](https://github.com/astro-informatics/sopt/runs/5886037546?check_suite_focus=true) - failing seed 1649425543 (ubuntu-20.04-g++-9-mpi:on-openmp:on )
- [ ] [test_primal_dual](https://github.com/astro-informatics/sopt/runs/6366759247?check_suite_focus=true) and [earlier](https://github.com/astro-informatics/sopt/runs/5883416496?check_suite_focus=true) - failing seed 1652173059 (ubuntu-20.04-g++-9-mpi:off-openmp:on ) and 1649413555 ( macos-11-clang++-mpi:on-openmp:off )
- [ ] [test_proximal](https://github.com/astro-informatics/sopt/runs/6366760329?check_suite_focus=true) and [earlier](https://github.com/astro-informatics/sopt/runs/5785291506?check_suite_focus=true) - failing seed 1652173370 (macos-11-clang++-mpi:off-openmp:on ) and 1648805187 ( macos-11-clang++-mpi:on-openmp:off )
- [x] Credible_region Already fixed in https://github.com/astro-informatics/sopt/pull/252

### CODE REVIEW

**Comment by SJaffa** [comment]
@CosmoMatt can you check that what I have tried here to avoid the failures makes sense?

In `sdmm` and `proximal` I increased `epsilon`, as if it is not quite converged then using a bigger step should find that it is close to the minima if not exactly there. If this doesn't work we can increase the iterations to get it to converge or investigate more.
In `primal_dual` I increased the margin on the `Approx()` check but it might need to be increased further.
In `power_method` we could change `power_tol` but that would influence a few other tests in that batch as well, is that ok?

If we fix the random seed I can use the seeds reported in the test failures to check if these edits worked, otherwise we can just wait and see if it fails again.

**Comment by SJaffa** [comment]
Interesting.. the tests all passed so I reran them and now 4 different tests failed! With the helpful error message `***Exception: Illegal`....
https://github.com/astro-informatics/sopt/runs/6919792311?check_suite_focus=true

Newly failing tests:
- [ ] test_example_inpainting
- [ ] test_example_padmm_inpainting
- [ ] test_example_pd_inpainting
- [ ] test_example_forward_backward_inpainting

I'm guessing the `pd` (primal dual?), and `padmm` (proximal_admm?) are unhappy with the changes I made, so we need to either check why they are unhappy and force them to be happy again or find a different way to avoid the random failures in the related tests. 
I'm not sure where the `test_example_inpainting` comes from, but if that's sdmm that would make sense. The forward_backward one shouldn't be having issues because I didn't change anything in the forward_backward test, but maybe it calls one of the others, will investigate. Also interesting why this didn't fail the first time.

**Comment by tkoskela** [comment]
## Some notes from today:

Catch tests get added by `add_catch_test` in [GreatCmakeCookoff](https://github.com/UCL/GreatCMakeCookOff/blob/master/scripts/AddCatchTest.cmake) that ultimately calls `add_catch_test_with_seed` but does not pass a seed, which defaults to the seed being set to `time`. You can pass a pair of `SEED <number>` to `add_catch_test` to seed the random seed in catch. Whether this gets actually used is uncertain, but looks that way since the random number generator is declared as `extern`.

In the examples they seem to create a new random number generator and seed it with `time(0)`. To seed them, you could just fix the seed for each rng (bad), or you could try to use the `extern` rng. Then you'd have to build the executables with the `common_catch_main_object`, not sure if that happens right now. The examples get added to catch by `AddExample.cmake` in `cmake_files`. Those call `add_test` directly which is a [cmake function](https://cmake.org/cmake/help/latest/command/add_test.html). It takes config options as argument and you can pass `--rng seed <number>` to set the random seed in the catch session configuration.

**Comment by SJaffa** [comment]
All the places where the mersenne twist random generator is seeded/used in the C++ code:
### std::mt19937 and std::srand seeded from time
- [./cpp/examples/proximal_admm/reweighted.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/proximal_admm/reweighted.cc#L53)
- [./cpp/examples/proximal_admm/inpainting.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/proximal_admm/inpainting.cc#L50) 
- [./cpp/examples/primal_dual/inpainting.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/primal_dual/inpainting.cc#L59) 
- [./cpp/examples/forward_backward/inpainting.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/forward_backward/inpainting.cc#L50) 
- [./cpp/examples/forward_backward/inpainting_credible_interval.cc](./cpp/examples/forward_backward/inpainting_credible_interval.cc)
- [./cpp/examples/forward_backward/inpainting_joint_map.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/forward_backward/inpainting_joint_map.cc#L51)
- [./cpp/examples/sdmm/reweighted.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/examples/sdmm/reweighted.cc#L54)
- [./cpp/examples/sdmm/inpainting.cc](./cpp/examples/sdmm/inpainting.cc)

### std::mt19937_64 
- [./cpp/tests/common_catch_main.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/common_catch_main.cc#L9) seeded 0 and [later in the same file](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/common_catch_main.cc#L17) seeded from catch `session.configData().rngSeed`
- [/cpp/tests/common_mpi_catch_main.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/common_mpi_catch_main.cc#L12) seeded 0 and [later](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/common_mpi_catch_main.cc#L46) seeded from catch `session.configData().rngSeed`

### from `extern`
- [./cpp/tests/sdmm.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/sdmm.cc#L11)
- [./cpp/tests/wavelets.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/wavelets.cc#L35) and [again the same](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/wavelets.cc#L40)
- -[./cpp/tests/forward_backward.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/forward_backward.cc#L18)
- [./cpp/tests/padmm.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/padmm.cc#L12)
- [./cpp/tests/proximal.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/proximal.cc#L12)
- [./cpp/tests/sara.cc](https://github.com/astro-informatics/sopt/blob/e6bcd564b82ca0d1cfea57f8da8aaeb88050b637/cpp/tests/sara.cc#L10)
- [./cpp/tests/serial_vs_parallel_padmm.cc](./cpp/tests/serial_vs_parallel_padmm.cc)

Other bits are also passed form CMake/catch2...

**Comment by SJaffa** [comment]
Looking at [this set of GHAction runs](https://github.com/astro-informatics/sopt/actions/runs/2543349228), passing the random seed through the  `add_catch_test` functions in `cpp/tests/CMakeLists.txt` seems to be working - they are failing where we expect and reporting the random seed I gave each test individually.

Except `proximal` which is now passing despite being passed the seed that failed before...

So now we can try changing the tolerance on the checks or similar until these pass.

File: cpp/tests/CMakeLists.txt [hunk 1]
```diff
@@ -15,25 +15,27 @@ if(CATCH_HAS_THROWS_AS)
...
+add_catch_test(sara LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(maths LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(wrapper LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(conjugate_gradient LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(linear_transform LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(chained_operators LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(sdmm LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(sdmm_warm_start LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test(proximal LIBRARIES sopt SEED ${RAND_SEED})
+add_catch_test_with_seed(seeded_proximal test_proximal ${RAND_SEED})
```
Comment (SJaffa) [inline cpp/tests/CMakeLists.txt]:
```suggestion
```
Can remove this now that the first test is seeded?

Comment (tkoskela) [reply cpp/tests/CMakeLists.txt]: Yes, this seems redundant now

**Review by tkoskela — APPROVED** [review]
Looks good, you can remove the duplicate seeded test now that all tests are seeded.

I have one comment about the tolerances which is a bit vague, sorry about that. I'll leave it to you to decide if this makes sense or not. I recall that we decided to raise the tolerances because the convergence criteria for the solver was lower than the tolerance and it was not reasonable numerically to expect the tolerance to be reached. Could we add an additional check that asserts the tolerance is higher than the convergence criteria, to justify the change and ensure it doesn't get broken again by some future change.

**Pull request merged**